File: PasswordGenerator/Python/Web/Flask/src/cache.py
import os
import json
import time
import threading

class CacheAdapter:
    def get(self, key: str):
        raise NotImplementedError
    def set(self, key: str, value, ttl: int = 300):
        raise NotImplementedError
    def delete(self, key: str):
        raise NotImplementedError
    def has(self, key: str) -> bool:
        raise NotImplementedError

class LocalCache(CacheAdapter):
    def __init__(self):
        self._store = {}
        self._lock = threading.Lock()
    def get(self, key: str):
        with self._lock:
            entry = self._store.get(key)
            if entry is None:
                return None
            if entry["expires"] is not None and time.time() > entry["expires"]:
                del self._store[key]
                return None
            return entry["value"]
    def set(self, key: str, value, ttl: int = 300):
        expires = time.time() + ttl if ttl else None
        with self._lock:
            self._store[key] = {"value": value, "expires": expires}
    def delete(self, key: str):
        with self._lock:
            self._store.pop(key, None)
    def has(self, key: str) -> bool:
        return self.get(key) is not None

class RedisCache(CacheAdapter):
    def __init__(self, url="redis://localhost:6379/0"):
        import redis as _redis
        self._client = _redis.from_url(url, decode_responses=True)
    def get(self, key: str):
        val = self._client.get(key)
        if val is None:
            return None
        return json.loads(val)
    def set(self, key: str, value, ttl: int = 300):
        self._client.setex(key, ttl, json.dumps(value))
    def delete(self, key: str):
        self._client.delete(key)
    def has(self, key: str) -> bool:
        return self._client.exists(key) > 0

def create_cache() -> CacheAdapter:
    cache_type = os.getenv("CACHE_TYPE", "redis").lower()
    if cache_type == "local":
        return LocalCache()
    try:
        url = os.getenv("REDIS_URL", "redis://localhost:6379/0")
        return RedisCache(url)
    except Exception:
        return LocalCache()

File: Inboxes/Node.js/Web/Express/src/cache.js
const os = require('os');

class CacheAdapter {
    get(key) { throw new Error('Not implemented'); }
    set(key, value, ttl) { throw new Error('Not implemented'); }
    delete(key) { throw new Error('Not implemented'); }
    has(key) { throw new Error('Not implemented'); }
}

class LocalCache extends CacheAdapter {
    constructor() {
        super();
        this._store = new Map();
    }
    get(key) {
        const entry = this._store.get(key);
        if (!entry) return null;
        if (entry.expires && Date.now() > entry.expires) {
            this._store.delete(key);
            return null;
        }
        return entry.value;
    }
    set(key, value, ttl = 300) {
        this._store.set(key, { value, expires: ttl ? Date.now() + ttl * 1000 : null });
    }
    delete(key) { this._store.delete(key); }
    has(key) { return this.get(key) !== null; }
}

class RedisCache extends CacheAdapter {
    constructor(url) {
        super();
        const Redis = require('ioredis');
        this._client = new Redis(url || 'redis://localhost:6379');
    }
    async get(key) {
        const val = await this._client.get(key);
        if (!val) return null;
        return JSON.parse(val);
    }
    async set(key, value, ttl = 300) {
        await this._client.setex(key, ttl, JSON.stringify(value));
    }
    async delete(key) { await this._client.del(key); }
    async has(key) { return (await this._client.exists(key)) > 0; }
}

function createCache() {
    const type = process.env.CACHE_TYPE || 'redis';
    if (type === 'local') return new LocalCache();
    try {
        const url = process.env.REDIS_URL || 'redis://localhost:6379';
        return new RedisCache(url);
    } catch (e) {
        return new LocalCache();
    }
}

module.exports = { createCache, LocalCache, RedisCache };
